Code/Framework/AzCore/Tests/AZStd/Hashed.cpp

```diff
@@ -287,6 +287,75 @@ namespace UnitTest
+            {
+                if (m_a == other.m_a)
+                {
+                    return m_b == other.m_b;
+                }
+                else if (m_a == other.m_b)
+                {
+                    return m_b == other.m_a;
+                }
+                return false;
+            }
+        };
+
+        // This hashing function produces different hashes for two equal values,
+        // which violates the requirement for hashing functions.
+        // The test would only make sure that this does not reproduce a bug that happened in the past in this case,
+        // where it would make the insert() to loop infinitely.
+        struct TwoPtrsHasher
+        {
+            size_t operator()(const TwoInts& p) const
+            {
+                size_t hash{ 0 };
+                AZStd::hash_combine(hash, p.m_a, p.m_b);
+                return hash;
+            }
+        };
+        using PairSet = AZStd::unordered_set<TwoInts, TwoPtrsHasher>;
+        PairSet set;
+        
+        set.insert({ 0x000001ceddd942a0, 0x000001ceddd94720 });
+        set.insert({ 0x000001ceddd94420, 0x000001ceddd94a20 });
+        set.insert({ 0x000001ceddd94720, 0x000001ceddd94420 });
+        set.insert({ 0x000001ceddd948a0, 0x000001ceddd945a0 });
+        set.insert({ 0x000001ceddd94a20, 0x000001ceddd9c420 });
+        set.insert({ 0x000001ceddd94ba0, 0x000001ceddd94d20 });
+        set.insert({ 0x000001ceddd94d20, 0x000001ceddd948a0 });
+        set.insert({ 0x000001ceddd9c2a0, 0x000001ceddd942a0 });
+        set.insert({ 0x000001ceddd9c420, 0x000001ceddd94ba0 });
+        set.insert({ 0x000001ceddd9c5a0, 0x000001ceddd9c720 });
+        set.insert({ 0x000001ceddd9c720, 0x000001ceddd9c8a0 });
+        set.insert({ 0x000001ceddd9c8a0, 0x000001ceddd94a20 });
+        set.insert({ 0x000001ceddd9ca20, 0x000001ceddd9cba0 });
+        set.insert({ 0x000001ceddd9cba0, 0x000001ceddd9cd20 });
+        set.insert({ 0x000001ceddd9cd20, 0x000001ceddd9cea0 });
+        set.insert({ 0x000001ceddd9cea0, 0x000001ceddd94a20 });
+        set.insert({ 0x000001ceddd9d020, 0x000001ceddd9e2c0 });
+        set.insert({ 0x000001ceddd9e2c0, 0x000001ceddd9e440 });
+        set.insert({ 0x000001ceddd9e440, 0x000001ceddd948a0 });
+        set.insert({ 0x000001ceddd9e5c0, 0x000001ceddd9e740 });
+        set.insert({ 0x000001ceddd9e740, 0x000001ceddd9e8c0 });
+        set.insert({ 0x000001ceddd9e8c0, 0x000001ceddd948a0 });
+        set.insert({ 0x000001ceddd9e5c0, 0x000001ceddd9d020 });
+        set.insert({ 0x000001ceddd9e440, 0x000001ceddd9cd20 });
+        set.insert({ 0x000001ceddd9d020, 0x000001ceddd9e440 });
+        set.insert({ 0x000001ceddd9cea0, 0x000001ceddd9e2c0 });
+        set.insert({ 0x000001ceddd9cd20, 0x000001ceddd9c720 });
+        set.insert({ 0x000001ceddd9cba0, 0x000001ceddd9ca20 });
+        // This line will toggle a rehash
+        set.insert({ 0x000001ceddd9ca20, 0x000001ceddd9cea0 });
```

Comment:
If I understood correctly, before the fix this unit test ended up in an infinite loop while rehashing, but now with the fix this should trigger an assert instead and delete the offensive element to avoid getting stuck. But I'm not seeing the unit test handling that an assert is expected, so this unit test should still fail due to the assert being triggered, right?

Reply:
Funny enough our unit tests don’t fail if there is an AZ_Assert on them. What would end up happening in the old scenario would be gtest timing out

Reply:
I just tested this and the tests do fail if an `AZ_Assert` is triggered, at least run from Visual Studio.

Reply:
You are absolutely right.
okay, so I found out that this only happens with void*, I changed it to size_t to be more readable but it didn't trigger the assert hence why I thought it was working fine.. will update, thanks!!

Reply:
Does this need a trace suppression assert around this block? As we do expect the assert to fire right?

Reply:
Yep, thats what I did in the latest version